## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics

df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Display the first 5 rows
df.head()
# Display the last 5 rows
# df.tail() 
# Display data types
# df.dtypes
# Display the row count and column count
# df.shape
# Display summary of the DataFrame, including data types and non-null counts
# df.info
# Display number of unique values in each column
# df.nunique()
# Checks for missing values in the entire DataFrame.
# df.isnull().sum()
# Diplay max value for each num column
# df.max()
# Diplay min value for each num column
#df.min()
# Display the mean value for each num column
#df.mean()










first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [5]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [30]:
# For the category with the most entries,
# Get the category with the most entries
most_entries_category = df['category'].value_counts().idxmax()
# most_entries_category

# Filter the dataframe for the category with the most entries
most_entries_category_df = df[df['category'] == most_entries_category]
# most_entries_category_df
# which subcategory had the most entries? Show the count.
most_entries_subcategory = most_entries_category_df['subcategory'].value_counts().idxmax()
most_entries_subcategory_count = most_entries_category_df['subcategory'].value_counts().max()
print(f'The subcategory with the most entries is {most_entries_subcategory} with {most_entries_subcategory_count} entries.')

The subcategory with the most entries is bathroom supplies with 6424 entries.


In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [31]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5)
client_ids = top_clients.index.tolist()
client_ids



[33615, 66037, 46820, 38378, 24741]

In [32]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_entries_client = df['client_id'].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == most_entries_client]['qty'].sum()
total_units_ordered


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [33]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping_price    total_price  line_cost  \
0                1  115164.00            75.0  125898.607500   80159.55   
1                0     523.95            14.9     588.693625     331.79   
2                6     527.28            16.8     594.407400     323.34   
3                3    1056.18            12.3    1167.314400     732.95   
4                1    3902.00           464.3    4770.182750    2627.70   
...            ...        ...             ...            ...        ...   
54634            8    2743.29            22.5    3021.625575    1725.30   
54635            0    9709.73           117.0   10735.702525    8363.62   
54636            6   31188.50            41.6   34118.884250   27263.85   
54637            0     165.76           180.4     378.179800     317.04   
54638            5    3103.31            20.7    3412.980925    1680.61   

        line_profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

In [34]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# Calculate the shipping price based on the weight of the items
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 if x > 50 else 10) * df['unit_weight']

# Display the updated dataframe
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping_price    total_price  line_cost  \
0                1  115164.00            75.0  125898.607500   80159.55   
1                0     523.95            14.9     588.693625     331.79   
2                6     527.28            16.8     594.407400     323.34   
3                3    1056.18            12.3    1167.314400     732.95   
4                1    3902.00           464.3    4770.182750    2627.70   
...            ...        ...             ...            ...        ...   
54634            8    2743.29            22.5    3021.625575    1725.30   
54635            0    9709.73           117.0   10735.702525    8363.62   
54636            6   31188.50            41.6   34118.884250   27263.85   
54637            0     165.76           180.4     378.179800     317.04   
54638            5    3103.31            20.7    3412.980925    1680.61   

        line_profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

In [35]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Calculate the total price including sales tax
df['total_price'] = df['subtotal'] + df['shipping_price'] + (df['subtotal'] + df['shipping_price']) * 0.0925

# Display the updated dataframe
df




first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping_price    total_price  line_cost  \
0                1  115164.00            75.0  125898.607500   80159.55   
1                0     523.95            14.9     588.693625     331.79   
2                6     527.28            16.8     594.407400     323.34   
3                3    1056.18            12.3    1167.314400     732.95   
4                1    3902.00           464.3    4770.182750    2627.70   
...            ...        ...             ...            ...        ...   
54634            8    2743.29            22.5    3021.625575    1725.30   
54635            0    9709.73           117.0   10735.702525    8363.62   
54636            6   31188.50            41.6   34118.884250   27263.85   
54637            0     165.76           180.4     378.179800     317.04   
54638            5    3103.31            20.7    3412.980925    1680.61   

        line_profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

# Display the updated dataframe
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  shipping_price  \
0             7.50   105            1  115164.00            75.0   
1             1.49    21            0     523.95            14.9   
2             1.68    39            6     527.28            16.8   
3             1.23    29            3    1056.18            12.3   
4            46.43    20            1    3902.00           464.3   
...            ...   ...          ...        ...             ...   
54634         2.25    33            8    2743.29            22.5   
54635        11.70    47            0    9709.73           117.0   
54636         4.16   475            6   31188.50            41.6   
54637        18.04   112            0     165.76           180.4   
54638         2.07  1031            5    3103.31            20.7   

         total_price  line_cost  
0      125898.607500   80159.55  
1         588.693625     331.79  
2         594.407400     323.34  
3        1167.314400     732.95  
4        4770.182750    2627.70  
...              ...        ...  
54634    3021.625575    1725.30  
54635   10735.702525    8363.62  
54636   34118.884250   27263.85  
54637     378.179800     317.04  
54638    

In [14]:
# Create a column for the profit of
# each line using line cost and line price
#df['line_profit'] = df['line_price'] - df['line_cost']

# Create a column for the profit of each line
df['line_profit'] = df['total_price'] - df['line_cost']


# Display the updated dataframe
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping_price    total_price  line_cost  \
0                1  115164.00            75.0  125898.607500   80159.55   
1                0     523.95            14.9     588.693625     331.79   
2                6     527.28            16.8     594.407400     323.34   
3                3    1056.18            12.3    1167.314400     732.95   
4                1    3902.00           464.3    4770.182750    2627.70   
...            ...        ...             ...            ...        ...   
54634            8    2743.29            22.5    3021.625575    1725.30   
54635            0    9709.73           117.0   10735.702525    8363.62   
54636            6   31188.50            41.6   34118.884250   27263.85   
54637            0     165.76           180.4     378.179800     317.04   
54638            5    3103.31            20.7    3412.980925    1680.61   

        line_profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [37]:
# Check your work using the totals above

# Filter the dataframe for the specified order IDs
order_ids = [2742071, 2173913, 6128929]
filtered_df = df[df['order_id'].isin(order_ids)]

# Calculate the total price for each order ID
total_prices = filtered_df.groupby('order_id')['total_price'].sum()
total_prices


# Compare the calculated total prices with the given totals
given_totals = [152811.89, 162388.71, 923441.25]
is_totals_match = total_prices.isin(given_totals)

is_totals_match


order_id
2173913    False
2742071    False
6128929    False
Name: total_price, dtype: bool

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [47]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

top_clients = df['client_id'].value_counts().head(5)
client_ids = top_clients.index.tolist()

client_ids

client_spending = df[df['client_id'].isin(client_ids)].groupby('client_id')['total_price'].sum()

# Sort the values in descending order

client_spending = client_spending.sort_values(ascending=False)
client_spending


# Format using Dollars

client_spending = client_spending.map('${:,.2f}'.format)
client_spending



client_id
24741    $76,678,479.05
38378     $9,172,866.14
66037     $8,746,255.31
46820     $8,006,332.53
33615     $6,392,918.35
Name: total_price, dtype: object

In [49]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Get the top 5 clients by quantity
top_clients = df['client_id'].value_counts().head(5)
client_ids = top_clients.index.tolist()

# Filter the dataframe for the top 5 clients
top_clients_df = df[df['client_id'].isin(client_ids)]

# Group the data by client_id and calculate the totals
summary_df = top_clients_df.groupby('client_id').agg({'qty': 'sum', 'shipping_price': 'sum', 'total_price': 'sum', 'line_profit': 'sum'})

# Sort the summary dataframe by total profit
summary_df = summary_df.sort_values('line_profit', ascending=False)

# Format the summary dataframe using Dollars

summary_df_formatted = summary_df.applymap('${:,.2f}'.format)
summary_df_formatted






/var/folders/d2/3s6fz4mj2jvcvytf_s89r5dh0000gn/T/ipykernel_31555/1153700291.py:21: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary_df_formatted = summary_df.applymap('${:,.2f}'.format)


qty shipping_price     total_price     line_profit
client_id                                                            
24741      $239,862.00      $9,365.60  $76,678,479.05  $36,106,662.11
66037       $43,018.00     $10,017.30   $8,746,255.31   $3,126,906.88
38378       $73,667.00     $11,895.00   $9,172,866.14   $2,955,705.56
46820       $75,768.00     $11,094.80   $8,006,332.53   $2,589,494.93
33615       $64,313.00     $12,609.40   $6,392,918.35   $2,033,979.93

In [50]:
# Format the data and rename the columns
# to names suitable for presentation.

df_formatted = df.rename(columns={'first':'First', 'last':'Last', 'job':'Job', 'phone':"Phone", 'email':'Email', 'client_id': 'Client ID', 
                                  'order_id': 'Order ID', 'order_date':'Order Date', 'order_week':'Order Week', 'order_year':'Order Year', 
                                  'item_id':'Item ID', 'category':'Category', 'subcategory':'Subcategory', 'item_name':'Item Name', 'qty':'Quantity',
                                  'line_number':'Line Number','unit_price':'Unit Price', 'unit_cost':'Unit Cost', 'unit_weight':'Unit Weight', 'subtotal':'Subtotal',
                                  'shipping_price':'Shipping Price', 'total_price':'Total Price', 'line_cost':'Cost', 'line_profit':'Profit'})

# Currency should be in millions of dollars.
# Format the data to reflect that.

df_formatted['Unit Price'] = df_formatted['Unit Price'].map('${:,.2f}'.format)
df_formatted['Unit Cost'] = df_formatted['Unit Cost'].map('${:,.2f}'.format)
df_formatted['Subtotal'] = df_formatted['Subtotal'].map('${:,.2f}'.format)
df_formatted['Shipping Price'] = df_formatted['Shipping Price'].map('${:,.2f}'.format)
df_formatted['Total Price'] = df_formatted['Total Price'].map('${:,.2f}'.format)
df_formatted['Cost'] = df_formatted['Cost'].map('${:,.2f}'.format)
df_formatted['Profit'] = df_formatted['Profit'].map('${:,.2f}'.format)

# Display the formatted dataframe
df_formatted








First       Last                      Job               Phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     Email  Client ID  Order ID  Order Date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       Order Week  Order Year  ... Unit Price Unit Cost Unit Weight Quantity  \
0              17        2023  ...  $1,096.80   $762.71        7.50      105   
1              20        2023  ...     $24.95    $15.09        1.49       21   
2               4        2023  ...     $13.52     $7.86        1.68       39   
3              17        2023  ...     $36.42    $24.85        1.23       29   
4              21        2023  ...    $195.10   $108.17       46.43       20   
...           ...         ...  ...        ...       ...         ...      ...   
54634          17        2023  ...     $83.13    $51.60        2.25       33   
54635           6        2023  ...    $206.59   $175.46       11.70       47   
54636           9        2023  ...     $65.66    $57.31        4.16      475   
54637           3        2023  ...      $1.48     $1.22       18.04      112   
54638           1        2023  ...      $3.01     $1.61        2.07     1031   

      Line Number     Subtotal  Shipping Price  Total Price        Cost  \
0               1  $115,164.00          $75.00  $125,898.61  $80,159.55   
1               0      $523.95          $14.90      $588.69     $331.79   
2               6      $527.28          $16.80      $594.41     $323.34   
3               3    $1,056.18          $12.30    $1,167.31     $732.95   
4               1    $3,902.00         $464.30    $4,770.18   $2,627.70   
...           ...          ...             ...          ...         ...   
54634           8    $2,743.29          $22.50    $3,021.63   $1,725.30   
54635           0    $9,709.73         $117.00   $10,735.70   $8,363.62   
54636           6   $31,188.50          $41.60   $34,118.88  $27,263.85   
54637           0      $165.76         $180.40      $378.18     $317.04   
54638           5    $3,103.31          $20.70    $3,412.98   $1,680.61   

           Profit  
0      $45,739.06  
1         $256.90  
2         $271.07  
3         $434.36  
4       $2,142.48  
...           ...  
54634   $1,296.33  
54635   $2,372.08  
54636   $6,855.03  
54637      $61.14  
54638   $1,732.37  

[54639 rows x 23 columns]

In [51]:
# Sort the updated data by "Total Profit" from highest to lowest

# Remove dollar signs and commas, then convert to float
df_formatted['Profit'] = df_formatted['Profit'].replace('[\$,]', '', regex=True).astype(float)

# Sort by 'Profit' in descending order
df_sorted = df_formatted.sort_values('Profit', ascending=False)
df_sorted['Profit'] = df_sorted['Profit'].map('${:,.2f}'.format)
df_sorted

First      Last                                    Job  \
21801     Julie  Anderson                    Exhibition designer   
15112    Dennis   Hawkins    Social research officer, government   
23929   Michele  Mcdonald  Emergency planning/management officer   
25972       Amy   Stewart              Psychologist, educational   
4063     Kendra   Garrett                 Pension scheme manager   
...         ...       ...                                    ...   
22529   Matthew    Orozco             Research scientist (maths)   
37936  Jennifer     Lewis                Agricultural consultant   
29972    Dennis   Hawkins    Social research officer, government   
21090    Daniel    Bowman                          Archaeologist   
38399     David     Green                      Personnel officer   

                       Phone                            Email  Client ID  \
21801             5933099462    anderson_better_4704@bell.biz      78965   
15112  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
23929     435-931-4939x95137    mcdonaldm317@hess-jackson.com      38473   
25972     979.793.4884x57205           stewartamy@francis.com      86721   
4063            808.257.2029  garrettkendra8479@jefferson.com      24741   
...                      ...                              ...        ...   
22529   +1-259-213-7449x4333           orozcomatthew@carr.com      67407   
37936             9813732574  lewis.jennifer@becker-ford.info      62566   
29972  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
21090   001-716-757-9724x862        danielbowman@thompson.com      72892   
38399     527.445.6304x24709    dgreen4039@sutton-leonard.org      57283   

       Order ID  Order Date  Order Week  Order Year  ... Unit Price Unit Cost  \
21801   3833568  2023-04-29          17        2023  ...    $147.66   $100.11   
15112   7263857  2023-03-05           9        2023  ...    $323.42   $183.84   
23929   4594154  2023-03-02           9        2023  ...    $674.94   $568.06   
25972   6057452  2023-03-21          12        2023  ...    $100.40    $73.05   
4063    5167984  2023-05-08          19        2023  ...    $461.30   $247.94   
...         ...         ...         ...         ...  ...        ...       ...   
22529   5546401  2023-05-18          20        2023  ...      $0.13     $0.12   
37936   5131256  2023-05-17          20        2023  ...      $0.13     $0.12   
29972   7438979  2023-03-27          13        2023  ...      $0.13     $0.12   
21090   6514417  2023-05-17          20        2023  ...     $96.86    $74.98   
38399   6619389  2023-05-22          21        2023  ...     $58.39    $49.69   

      Unit Weight Quantity Line Number         Subtotal  Shipping Price  \
21801        0.09  3958244           1  $584,474,309.04           $0.90   
15112        2.21   637318           3  $206,121,387.56          $22.10   
23929        2.25   219645           5  $148,247,196.30          $22.50   
25972        9.22   977121           3   $98,102,948.40          $92.20   
4063         1.86   122166           7   $56,355,175.80          $18.60   
...           ...      ...         ...              ...             ...   
22529        0.28        8           0            $1.04           $2.80   
37936        0.28        6           2            $0.78           $2.80   
29972        0.28        1           6            $0.13           $2.80   
21090        0.24        0           6            $0.00           $2.40   
38399        0.18        0           0            $0.00           $1.80   

           Total Price             Cost           Profit  
21801  $638,538,183.61  $396,259,807.74  $242,278,375.87  
15112  $225,187,640.05  $117,164,563.22  $108,023,076.83  
23929  $161,960,086.54  $124,771,561.20   $37,188,525.34  
25972  $107,177,571.86   $71,378,781.25   $35,798,790.61  
4063    $61,568,049.88   $30,289,856.64   $31,278,193.24  
...                ...              ...             